src/loaders/tiff/lib/utils.ts

```diff
@@ -0,0 +1,98 @@
+import type { GeoTIFF } from 'geotiff';
+import type { OmeTiffSelection } from '../ome-tiff';
+import type { OMEXML } from '../../omexml';
+import { getDims } from '../../zarr/lib/utils';
+
+function getIFDIndexer(imgMeta: OMEXML[0]): (sel: OmeTiffSelection) => number {
+  const { SizeC, SizeZ, SizeT, DimensionOrder } = imgMeta.Pixels;
+  switch (DimensionOrder) {
+    case 'XYZCT': {
+      return ({ t, c, z }) => t * SizeZ * SizeC + c * SizeZ + z;
+    }
+    case 'XYZTC': {
+      return ({ t, c, z }) => c * SizeZ * SizeT + t * SizeZ + z;
+    }
+    case 'XYCTZ': {
+      return ({ t, c, z }) => z * SizeC * SizeT + t * SizeC + c;
+    }
+    case 'XYCZT': {
+      return ({ t, c, z }) => t * SizeC * SizeZ + z * SizeC + c;
+    }
+    case 'XYTCZ': {
+      return ({ t, c, z }) => z * SizeT * SizeC + c * SizeT + t;
+    }
+    case 'XYTZC': {
+      return ({ t, c, z }) => c * SizeT * SizeZ + z * SizeT + t;
+    }
+    default: {
+      throw new Error(
+        `Invalid DimensionOrder in OME-XML, got ${JSON.stringify(
+          DimensionOrder
+        )}.`
+      );
+    }
+  }
+}
+
+// TODO: Indexers assume first image in metadata. Need to compute offsets differently
+// for other images in the OME-XML.
+export function getLegacyIndexer(rootMeta: OMEXML) {
+  const imgMeta = rootMeta[0];
+  const ifdIndexer = getIFDIndexer(imgMeta);
+
+  return async (sel: OmeTiffSelection, pyramidLevel: number) => {
+    const imgOffset = ifdIndexer(sel);
+    const { SizeT, SizeC, SizeZ } = imgMeta.Pixels;
+    const pyramidOffset = pyramidLevel * SizeZ * SizeT * SizeC;
+    return imgOffset + pyramidOffset;
+  };
+}
+
+export function getSubIFDIndexer(tiff: GeoTIFF, rootMeta: OMEXML) {
+  const imgMeta = rootMeta[0];
+  const ifdIndexer = getIFDIndexer(imgMeta);
+
+  return async (sel: OmeTiffSelection, pyramidLevel: number) => {
+    const offset = ifdIndexer(sel);
+
+    // It's the highest resolution, no need to look up SubIFDs.
+    if (pyramidLevel === 0) return offset;
+
+    // Lookup SubIFDs from the base image, and return offset to selection.
+    const baseImage = await tiff.getImage(offset);
+    const { SubIFDs } = baseImage.fileDirectory;
+    if (!SubIFDs) {
+      throw Error('Indexing Error: OME-TIFF is missing SubIFDs.');
+    }
+    return SubIFDs[pyramidLevel - 1];
+  };
+}
```

Comment:
Previously, when an image has subIFDs, we computed the image index as:

```javascript
const index = this._getIFDIndex(sel);
const pyramidIndex = index + z * SizeZ * SizeT * SizeC;

const parentImage = await this.tiff.getImage(index);
const { SubIFDs } = parentImage.fileDirectory;
this.tiff.ifdRequests[pyramidIndex] = this.tiff.parseFileDectoryAt(SubIFDs[z - 1]);

const image = await tiff.getImage(pyramidIndex); // pyramidIndex could be _anything_ and this would work
```

which means we were taking the filedirectory at index `SubIFDs[z - 1]` and just manually adding it to whatever index `pyramidIndex` is. What we really want to do is return the right `SufIFD` from the indexer.